Update 9-design.md (#9)
swent-sp-2024-event-radar/mvp-prd: template/9-design.md

# Design and Implementation

## Frontend
### Implementation Framework

> The app will be developed in Kotlin, optimized for Android, utilizing the Model-View-ViewModel (MVVM)architecture. This approach ensures a clear separation of concerns, making the management of the app’s UI and business logic more efficient and maintainable.

Reviewer: ```suggestion
The app will be developed in Kotlin, optimized for Android, utilizing the Model-View-ViewModel (MVVM)^architecture. This approach ensures a clear separation of concerns, making the management of the app’s UI and business logic more efficient and maintainable.
```

### Key Libraries and Languages
- **Kotlin**: Primary language for our Android development
- **Hilt**: Dependency injection library for managing dependencies and improving code modularity.
- **Material3**: Used for all graphic components
- **Jetpack Compose**: Used for dynamic UI rendering.
- **Firebase**: Backend-as-a-Service platform providing authentication, database, and cloud storage.
- **Android Camera Image Analysis**: Utilized for integrating camera functionalities required for photo capture during friend requests.
- **Dagger Assisted Factory**: Design pattern to create objects, particularly useful in the ViewModel creation process, ensuring that instances are correctly initialized with necessary dependencies.

> - **ZXing**: Library to encode information to QR codes and decode QR codes to information

Reviewer: Not sure if we should add Nominatim API for location searching?

> - **ZXing**: Library to encode information to QR codes and decode QR codes to information

Reviewer: also need to add something about 3rd party payment api?

Author: You're right! Added a third-party payment SDK to integrate this

### Dynamic UI Rendering Strategy
The app will leverage Jetpack Compose for dynamic UI rendering. Compose enables the creation of UI components declaratively, allowing for real-time updates and modifications based on data changes.
- **Composable Functions**: These will be used to define each screen and UI component, enabling reusable and modular UI elements.
- **State Management**: ViewModels will manage the state, ensuring the UI is always in sync with the underlying data model.
### Essential Screens
#### 1 Home Screen
- **Features**: Displays all events as well as events the user has registered to in either list or map view

> - **Components**: Event list, Upcoming events, search-filter, list/map view button

Reviewer: perhaps mention nav bar?

Author: Added it as an extra Essential Screen and explained that it was at the bottom of all other screens

##### 2.1 Event Creation Screen
- **Features**: Enables users to create new events with detailed specifications
- **Components**: Input Fields for Event Details (Name, Photo, Description, Category, Start/End Date & Time, Location, Ticket Info), and a create event button
- **Description**: The Event Creation Screen simplifies the event creation process, providing users with input fields for all necessary event details. The interface encourages users to organize and host events.
#### 3 User Profile Screen
- **Features**: Displays the user's personal information for viewing and editing.
- **Components**: User's Personal Information Input Fields (Text fields or input fields, depending on edit mode) and edit button
- **Description**: The User Profile Screen offers a view of the user's information, allowing easy access for review and modification. 
#### 4 Messages Screen
- **Features**: Facilitates real-time messaging between users.

> - **Components**: Chat list and My Friends’ chat list

Reviewer: ```suggestion
- **Components**: Chat list and My Friends list
```

> - **Description**: The Messaging Screen serves as a central hub for users to engage in real-time conversations. It provides a comprehensive view of all conversations, allowing users to seamlessly switch between chats. Additionally, a tab allows users to specifically view conversations with their friends, streamlining communication with close contacts.

Reviewer: ```suggestion
- **Description**: The Messaging Screen serves as a central hub for users to engage in real-time conversations. It provides a comprehensive view of all conversations, allowing users to seamlessly switch between chats. Additionally, a tab allows users to specifically view their friends list, streamlining the ability to review the connections made and communications with close contacts.
```

#### 6 Event Details Screen
- **Features**: Displays detailed information about a selected event, including date, time, location, and group chat (not in POC)
- **Components**: Event description, group chat redirection button
- **Description**: The Event Details Screen offers comprehensive details about the selected event, enhancing the user's understanding of the event's key aspects. Additionally, registered users can access the event's group chat (not available in POC), facilitating communication and engagement among attendees.
### Backend Communication
The app will utilize Firebase for backend communication, leveraging its robust set of tools and services:
- **Authentication**: Firebase Authentication for secure login and user management.
- **Database**: Firebase Firestore for real-time data synchronization and storage.
- **Cloud Storage**: Firebase Storage for securely storing user-uploaded photos and other media.

## Backend
### User Management
- **Description**: This functional block handles user-related operations such as registrations, profiles, and interactions between users (e.g., becoming friends).

> - **Application Logic**: Manages user authentication, registration, profile updates, and friend requests.

Reviewer: ```suggestion
- **Application Logic**: Manages user authentication, registration, profile updates, and adding friends via QR scan.
```

- **Framework**: Developed using Firebase Authentication and Firestore for data storage.
- **Dynamic View Rendering**: Integrates with the frontend to render user profiles and friend lists dynamically based on data from Firestore.
- **Responsive Backend-Frontend Interaction**: Ensures responsive interaction with the frontend, updating user profiles and friend lists based on user interactions.
### Event Management
- **Description**: This block handles operations related to creating, updating, and managing events, including ticket management.
- **Application Logic**: Manages event creation, updates, and ticket sales.
- **Framework**: Utilizes Firebase Firestore for event data storage.

> - **Dynamic View Rendering**: Provides APIs to fetch event details dynamically, ensuring frontend displays reflect the latest event information.

Reviewer: add something about payment API?

Author: Modified as well to take that into account

- **Responsive Backend-Frontend Interaction**: Updates event details on the frontend based on user actions such as event creation or ticket purchases.
### Messaging System
- **Description**: This block manages communication between users through real-time messaging.
- **Application Logic**: Handles message sending and retrieval between users.
- **Framework**: Utilizes Firebase Firestore for real-time messaging.
- **Dynamic View Rendering**: Updates message threads dynamically based on new messages received.
- **Responsive Backend-Frontend Interaction**: Ensures real-time updates of message threads on the frontend as new messages are sent or received.
### Additional Considerations
- **Monitoring and Logging**: Utilize Firebase Analytics and Firebase Crashlytics to track backend performance and identify issues promptly.
- **Scalability**: Design the backend to scale horizontally to handle increasing user loads effectively using Firebase's scalability features.

## Data Model
### Data Collection and Management
#### User Data

> - **Names:** User's first and last names. (String)

Reviewer 2: Nit: Maybe we could change all of these from plural to singular: e.g. Names to name, email addresses to email address, etc to ensure consistency.

Author: You're right! Everything was corrected

- **Email Addresses:** User's email for authentication and communication. (String)
- **Friends:** List of friends connected within the app. (Array of userIds)

> - **Bio:** Short presentation text. (String)

Reviewer 2: ```suggestion
- **Bio:** Short description of a user. (String)
```

- **Events Host List:** List of events hosted by the user. (Array of eventIds)
- **QR Code URL:** URL of the user's QR code image. (String - URL)
#### Event Data
- **Name**: Name of the event (String)
- **Photo**: Link to the photo stored of the event (String - URL)
- **Description**: Short text presenting the event (String)
- **Categories**: Categories or tags describing the type of event. (Array of Strings)
- **Start and End Date**: Date and time when events begin and end. (String - Date)
- **Location**: Details of the location
- **Location name**: Name of the location (String)
  - **Location lat**: Latitude of the location (Number)
  - **Location lng**: Longitude of the location (Number)
- **Ticket Information**: Details of the ticket
  - **Ticket name**: Name of the type of the ticket (String)
  - **Ticket price**: Price of the type of the ticket (Number)
  - **Ticket capacity**: Number of available tickets of the type of the ticket (Number)
  - **Ticket purchases**: Number of reserved tickets of the type of the ticket (Number)
- **Attendee List**: List of users participating in events. (Array of userIds)
- **Main Organizer**: Id of the user organizing the event. (String)

> - **Organizer List**: List of users organizing events. (Array of userIds)

Reviewer 2: ```suggestion
- **Organizer List**: List of user ids of users organizing events. (Array of userIds)
```

- The application is developed using modern development practices and frameworks such as Kotlin for Android development.
- Comprehensive testing methodologies, including unit testing, integration testing, and end-to-end testing, are employed to ensure the reliability and functionality of the application.
- Continuous Integration/Continuous Deployment (CI/CD) pipelines are set up to automate the testing and deployment processes, allowing for rapid iterations and updates.
### Deployment
- The application is deployed on cloud platforms like AWS or Azure to ensure high availability and scalability.
- Special attention is paid to infrastructure requirements to support the expected load and user base of the application.
- Firebase Authentication is utilized to secure APIs, ensuring that data requests are authenticated and users can only access data relevant to them.
- HTTPS is enforced for all communications between the app and servers to guarantee the confidentiality and integrity of data transmission.
### Special Infrastructure Requirements
- Scalable backend infrastructure is provisioned to handle potential spikes in traffic, especially during peak usage periods or major events.

> - Load balancing mechanisms are implemented to distribute traffic evenly across multiple server instances, optimizing performance and reliability.

Reviewer 2: Do we need load balancing for the MVP initially? Maybe this is when we have a larger user base.

Author: I have taken it out as there shouldn't too much issues in the beginning indeed

- Databases are configured for high availability and durability, with appropriate backup and recovery mechanisms in place to safeguard against data loss.
- Monitoring and logging systems are set up to track system performance, detect anomalies, and troubleshoot issues in real-time, ensuring the smooth operation of the application.

## Test Plan
### Testing Strategies
#### UI Tests
- UI tests are conducted to verify that UI components, including those instantiated by factories, run smoothly and interact correctly.
- Tests cover UI elements such as buttons, input fields, dialogs, etc., instantiated by factories.
#### ViewModel Tests
- ViewModel tests ensure that appropriate states are prepared for UI components instantiated by factories.

> - Tests verify the logic and behavior of ViewModel classes, including interactions with factory-created UI components.

Reviewer 2: ```suggestion
- Tests verify the logic and behavior of ViewModel classes, including interactions with factory-created UI components.
- Tests utilize Mock Repositories for the databases to isolate testing specifically on the ViewModel.
```
